src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,321 @@
+        schedule(nextSendEvent, curTick() + 100);
+    }
+}
+
+
+void
+GUPSGen::sendNextBatch()
+{
+    if (port.blocked()){
+        return;
+    }
+
+    if (!requestPool.empty())
+    {
+        PacketPtr pkt = requestPool.front();
+
+        exitTimes[pkt] = curTick();
+        DPRINTF(GUPSGen, "sendNextBatch: Sent pkt, pkt->addr_range: %s,"
+                        "(pkt != nullptr): %d\n",
+                        pkt->getAddrRange().to_string(), pkt != nullptr);
+        port.sendTimingPacket(pkt);
+        onTheFlyRequests++;
+        requestPool.pop();
+    }
+
+    if (!nextSendEvent.scheduled() && !port.blocked()
+                                && !requestPool.empty()) {
+        schedule(nextSendEvent, curTick() + 250);
+    }
+
+    // if (requestPool.empty())
+    // {
+    //     generateNextBatch();
+    // }
+
+    return;
+}
+
+PacketPtr
+GUPSGen::getReadPacket(Addr addr, unsigned int size)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size,
+                                            0, requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::ReadReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
+    pkt->dataDynamic(pkt_data);
+
+    return pkt;
+}
+
+PacketPtr
+GUPSGen::getWritePacket(Addr addr, unsigned int size, uint8_t *data)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size, 0,
+                                               requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::WriteReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
+    pkt->dataDynamic(pkt_data);
+
+    // std::fill_n(pkt_data, req->getSize(), (uint8_t)requestorId);
+    std::memcpy(pkt_data, data, req->getSize());
+
+    return pkt;
+}
+
+
+uint64_t
+GUPSGen::randomNumberGenerator(int64_t n)
+{
+    int i, j;
+    uint64_t m2[64];
+    uint64_t temp, ran;
+
+    while (n < 0) n += PERIOD;
+    while (n > PERIOD) n -= PERIOD;
+    if (n == 0) return 0x1;
+
+    temp = 0x1;
+    for (i = 0; i < 64; i++)
+    {
+        m2[i] = temp;
+        temp = (temp << 1) ^ ((int64_t) temp < 0 ? POLY : 0);
+        temp = (temp << 1) ^ ((int64_t) temp < 0 ? POLY : 0);
+    }
+
+    for (i = 62; i >= 0; i--)
+    {
+        if ((n >> i) & 1)
+            break;
+    }
+
+    ran = 0x2;
+    while (i > 0)
+    {
+        temp = 0;
+        for (j = 0; j < 64; j++)
+            if ((ran >> j) & 1)
+            temp ^= m2[j];
+        ran = temp;
+        i -= 1;
+        if ((n >> i) & 1)
+            ran = (ran << 1) ^ ((int64_t) ran < 0 ? POLY : 0);
+    }
+
+    return ran;
+}
+
+Addr
+GUPSGen::indexToAddr (uint64_t index)
+{
+    Addr ret = index * 8 + startAddr;
```

Comment:
what is `8`?

Reply:
Each element in the array in the code is uint64_t which is 8 bytes

Reply:
You should (almost) never use magic numbers like this. Instead, use `sizeof(uint64_t)`. This way, if you ever change this type you (might) remember to change this. But, more importantly, it makes the code more readable.

BTW, when I ask these questions I'm looking for you to update the code to answer them, most of the time.